src/analyses/call_graph_helpers.cpp

```
@@ -102,6 +102,15 @@ std::set<irep_idt> get_functions_reachable_within_n_steps(
     graph, start_functions, n, forwards);
 }
 
+void disconnect_unreachable_functions(
+  call_grapht::directed_grapht &graph,
+  const irep_idt &function)
+{
+  std::size_t start_idx = *(graph.get_node_index(function));
+  if(start_idx != optionalt<std::size_t>())
```

If it was Empty like this the deref on the line above would already have faulted. This should probably be
```
if(auto start_idx = graph.get_node_index(function))
  graph.disconnect_unreachable(*start_idx);
```

(This line compiled because `optionalt<std::size_t>` has an implicit conversion from `std::size_t`, so the current test is equivalent to `if(optionalt<std::size_t>(start_idx) != optionalt<std::size_t>())`, which is always true)

---

There don't actually seem to be any checks that optional arguments have been returned elsewhere in this file, e.g., 
` graph.get_reachable(*(graph.get_node_index(function)), forwards);` is used in `get_connected_functions`. 

Do I need the check? What happens in the above case if the function isn't found and therefore no argument is returned?

---

The `*` operator would cause a failing assertion. If you want to do a check, don't apply `operator*` and test `.has_value()` of the result before using (i.e., dereferencing).

---

So, if I want to throw an error anyway, I may as well just do what is done in the get_reachable function?

---

If you're happy with an assertion failure then yes simply remove the `if(start_idx != optionalt<std::size_t>())`. Since it's using old-fashioned `assert` the result will be an assert failure on a debug build, or continuing with undefined behaviour on a release build.